<a href="https://colab.research.google.com/github/ElizangelaStudent/L-GICA-DE-PROGRAMA-O-I-PY-/blob/main/Aula_7_Strings_em_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**String em Python**

###Uma string é uma coleção de caracteres - como se fosse uma lista, mas aceitando apenas elementos textuais. 

###Mas antes de adentramos na string, vamos conhecer uma função importante para nosso aprendizado:

##**Função Help**


>>A função python help() invoca o sistema de ajuda integrado interativo. Se o argumento for uma sequência, a sequência será tratada como o nome de um módulo, função, classe, palavra-chave ou tópico de documentação, e uma página de ajuda é impressa no console. Se o argumento for qualquer outro tipo de objeto, uma página de ajuda no objeto será exibida.


###**Keywords**

>**Keywords**: são palavras-chave especiais reservadas que têm significados e propósitos específicos e não podem ser usadas para nada além desses propósitos específicos. Essas palavras-chave estão sempre disponíveis — você nunca terá que importá-las em seu código.

>>Um exemplo de algo que você não pode fazer com palavras-chave Python é atribuir algo a eles. Se você tentar, então você vai ter uma SintaxeError. Você não terá um SintaxeError se tentar atribuir algo a uma função ou tipo embutido, mas ainda não é uma boa ideia. Para obter uma explicação mais aprofundada de como as palavras-chave podem ser mal utilizadas, confira a Sintaxe Inválida em Python: Common Reasons for SyntaxError.



In [ ]:
help('keywords') 


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 



#####Podemos realizar essa função com a palavra str, veja abaixo:

In [ ]:
help('str')

Help on class str in module builtins:

class str(object)
 |  str(object='') -> str
 |  str(bytes_or_buffer[, encoding[, errors]]) -> str
 |  
 |  Create a new string object from the given object. If encoding or
 |  errors is specified, then the object must expose a data buffer
 |  that will be decoded using the given encoding and error handler.
 |  Otherwise, returns the result of object.__str__() (if defined)
 |  or repr(object).
 |  encoding defaults to sys.getdefaultencoding().
 |  errors defaults to 'strict'.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __format__(self, format_spec, /)
 |      Return a formatted version of the string as described by format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  

##**String em Python**

###Um tipo de dados bastante usado no dia a dia são as strings, ou cadeias de caracteres (ou sequências de caracteres). O tipo de dados string, ou **str** como é chamado em Python, possui várias operações úteis associadas a ele. Essas operações tornam Python uma linguagem bastante propícia para manipulação de textos.

###Os exemplos abaixo ilustram o tipo de dados string e as operações associadas a ele.

##**Entrada e saida**

- Entrada: input
- Saida: print

In [ ]:
string  = input('Digite uma mensagem:')
print(string)

Digite uma mensagem:Boa noite!!
Boa noite!!


###**str.format()** é um dos métodos de formatação de string em Python3, que permite várias substituições e formatação de valor. Esse método nos permite concatenar elementos em uma string por meio da formatação posicional.

###Usando um único formatador:
###Os formatadores trabalham colocando um ou mais campos de substituição e espaços reservados definidos por um par de chaves {} em uma string e chamando str.format(). O valor que desejamos colocar nos espaços reservados e concatenar com a string passada como parâmetros para a função de formato.

###Sintaxe: {} .format (valor)

###Parâmetros:
- (valor): Pode ser um número inteiro, constante numérica de ponto flutuante, string, caracteres ou mesmo variáveis.

- Returntype: Retorna uma string formatada com o valor passado como parâmetro na posição do marcador.

In [ ]:
pi = 3.14
print('O valor do pi é {}.'.format(pi))   #utilizando o format

O valor do pi é 3.14.


In [ ]:
a = 10
b = '20'
c = 30.5

print('valores %d %s %.2f' % (a, b, c))   # essa 2.f representa o numero float e a quantidade de casas que a pessoa deseja, nesse caso são duas.

valores 10 20 30.50


##**Conversão de valores**

###A conversão se dará por meio das operações aritmeticas vista nas aulas anteriores

In [ ]:
30/10

3.0

##**Tipo numérico para string**

###**Tipo inteiro**

In [ ]:
pop_string = '200_000_000'    #utilizando a palavra int
pop_inteiro = int(pop_string)

print(f'A população brasileira é de ' + str(pop_inteiro) + ' pessoas.')
type(pop_inteiro)

###**Tipo real**

In [ ]:
pi_float = 3.14             #utilizando a palavra float
pi_string = str(pi_float)

print(f'O tipo de pi_float é {type(pi_float)} e o tipo de pi_string é {type(pi_string)}' )

O tipo de pi_float é <class 'float'> e o tipo de pi_string é <class 'str'>


##**Teste 1**

**Faça um programa que execute a lógica de login em um sistema. <br>
O programa deve solicitar que o usuário digite um `nome de usuário` e `senha`.<br> 
O programa deve verificar se o `nome de usuário` é igual a "admin" e `senha` =  "admin123" . <br>**
> Se os dados estiverem corretos, o programa deve exibir uma mensagem de boas-vindas.<br> Caso contrário, exibir uma mensagem de erro.

In [ ]:
while True:
    user = input('Digite o usuário: ')
    senha = input('Digite a senha: ')

    if (user != 'admin') or (senha != 'admin123'):
        print(f'Seu {user} não foi encontrado.')
    else:
        print(f'Bem vindo {user}.')
        break

Digite o usuário: ADMIN
Digite a senha: ADMIN123
Seu ADMIN não foi encontrado.
Digite o usuário: admin
Digite a senha: admin123
Bem vindo admin.


##**Criando strings**

In [ ]:
string1 = 'Vamo codar'
string2 = "Let's code"

In [ ]:
string1 = 'Vamo codar'
string2 = "string 1 \
           string 2 \
           string 3"
string2

'string 1            string 2            string 3'

In [ ]:
string1 = '''Vamo codar'''
string2 = """Let's Code"""

##**Acessando os elementos de uma string**

>podemos acessar por meio dos indices, temos dois tipos:
 - **reverso**: negativo e decrescente
 - **progressivo**: positivo e crescente

![title](https://s3-sa-east-1.amazonaws.com/lcpi/37773811-c38d-496c-a810-c73fe072b684.png)

In [ ]:
l= [4,6,8,9]
l[-2]


8

In [ ]:
l[2]

8

##**Acessando partes de uma string**

> **string[ início : parada]** <br>
    - início: incluído <br>
    - parada: não incluído <br>

###**Teste 2**

###Palíndromos são palavras que possuem a mesma sequencia de letras tanto da esquerda para a direita quanto da direita para a esquerda. Faça um programa que receba do usuário uma entrada qualquer e verifique se a entrada é um palíndromo.

In [ ]:
palavra = input("Digite uma palavra: ")
if str(palavra) == str(palavra)[::-1] :
    print("Essa palavra é um Palindrome")
else:
    print("Essa palavra não é Palindrome")

Digite uma palavra: ana
Essa palavra é um Palindrome


##**Strings vs Listas**

###Dizemos que uma string é imutável. Isso significa que não podemos alterar um elemento individual da string. O programa abaixo produz um erro:



In [ ]:
frase = "let's code"
frase[0] 


'l'

>Para alterar uma string, é necessário redefini-la, de modo que a string original será descartada e a nova (alterada) será escrita por cima da original. Ou, alternativamente, podemos gerar uma cópia da string com alterações. Veremos mais detalhes adiante.

###As Strings como ditas anteriores, elas são imutáveis, e isso pode nos dar um pouco de trabalho quando queremos fazer pequenas alterações, como forçar um caractere a ser maiúsculo ou acrescentar um caractere à string. Uma das formas de fazer envolve a "soma cumulativa" apresentada acima. Outra forma envolve **transformar a nossa string em lista**, que é uma estrutura mutável. Execute o programa abaixo:

In [ ]:
string = "let's Code"
lista = list(string)
lista[0] = 'L'
lista.append('!')
print(lista)
# resultado:
# ['L', 'e', 't', "'", 's', ' ', 'C', 'o', 'd', 'e', '!']

['L', 'e', 't', "'", 's', ' ', 'C', 'o', 'd', 'e', '!']


###Como a lista é mutável, nela conseguimos alterar uma letra e adicionar um símbolo ao final sem dificuldades! Porém, infelizmente nosso resultado é uma lista, o que não ficou muito legível para o usuário. Podemos resolver isso utilizando a função join. Veremos em breve como ele realmente funciona, mas por hora podemos utilizá-lo da seguinte maneira para transformar lista em string:

In [ ]:
string_original = "let's Code"
lista = list(string_original)
lista[0] = 'L'
lista.append('!')
string_final = ''.join(lista) # antes do . temos uma string vazia
print(string_final)
# resultado: "Let's Code!"
##Para as modificações mais comuns, temos algumas funções prontas que poderão ser bastante úteis!

Let's Code!


##**Operações com strings**

###Alguns operadores aritméticos funcionam com strings também. Naturalmente, eles não servem para fazer contas, mas nos permitem fazer de forma intuitiva algumas operações bastante úteis.

###O operador + serve como um operador de concatenação de strings: unir duas strings. Observe o exemplo abaixo:



In [ ]:
string1 = 'Olá'
string2 = 'Mundo'
resultado = string1 + string2  
print(resultado) # Na tela: 'OláMundo'


OláMundo


**para colocar as palavras separadas podemos acrescentar o operador aritmetico mais e aspas vazias(+" "), veja abaixo:** 

In [ ]:
string1 = 'Olá'
string2 = 'Mundo'
resultado = string1 + " "+ string2  
print(resultado) # Na tela: 'Olá Mundo'

Olá Mundo


###Outro operador que funciona é o operador *. Este operador não é usado entre duas strings, mas entre uma string e um int. Ele repetirá a string o número de vezes dado pelo int.

In [ ]:
string = 'Olá mundo ! '
multi = string * 3
print(multi) # Na tela: 'Olá mundo ! Olá mundo ! Olá mundo !'

Olá mundo ! Olá mundo ! Olá mundo ! 


##**Operações lógicas**

###Os operadores lógicos (<, >, <=, =>, != e ==) também funcionam com strings. Esses operadores são case sensitive, ou seja, diferenciam maiúsculas de minúsculas.

###De maneira muito simplificada e desconsiderando diferenças entre maiúsculas e minúsculas, podemos dizer que eles consideram ordem alfabética: 'banana' é maior do que 'abacaxi'.

###A explicação mais completa é a seguinte: internamente, cada caractere é armazenado como um número. Quando utilizamos qualquer tipo de aplicação que irá exibir um texto (um editor de textos, navegador de internet, ou mesmo os nossos programinhas em Python rodando no terminal), a aplicação usa esses números como índices em uma tabela de codificação de caracteres.

##**Tabela ASCII e outras representações**

![title](https://s3-sa-east-1.amazonaws.com/lcpi/f2abe68f-4eb7-4774-9438-d622328ea125.gif)

###A explicação mais completa é a seguinte: internamente, cada caractere é armazenado como um número. Quando utilizamos qualquer tipo de aplicação que irá exibir um texto (um editor de textos, navegador de internet, ou mesmo os nossos programinhas em Python rodando no terminal), a aplicação usa esses números como índices em uma tabela de codificação de caracteres.

###Temos diversos esquemas diferentes de codificação de caracteres em uso pelo mundo, e quando você está usando um programa ou navegando por um site e você nota símbolos estranhos no texto (frequentemente onde teríamos caracteres especiais, como letras com acento), é provável que o autor do texto tenha utilizado uma tabela e o seu computador esteja usando outra.

###Vários programas permitem a conversão entre essas tabelas, e você já deve ter visto essa "sopa de letrinhas" em alguma aba ou janela de configuração em algum editor de textos: utf-8, utf-16, windows-1252 (ou cp-2152), e até mesmo alguns padrões ISO.

###Para ilustrar a ideia, vamos colocar aqui uma das tabelas mais simples e abaixo dele tem o codigo que mostra cada item da tabela ASCII:

![title2](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1b/ASCII-Table-wide.svg/1200px-ASCII-Table-wide.svg.png)

###Note que o caractere 'A' está no índice 65, o 'B' está no índice 66, e assim sucessivamente. Por isso, 'Abacate' < 'Banana' é verdadeiro: a primeira string começa com uma letra no índice 65 da tabela, a segunda com uma letra no índice 66.

>No caso de 'Abacate' e 'Abacaxi', temos um "empate" das 5 primeiras letras. Então é a sexta letra que vai mandar: 'x' é maior do que 't', por estar em uma posição superior na tabela.

###Note que a ordem não é exatamente alfabética: entre os caracteres maiúsculos, seguimos ordem alfabética. Entre os minúsculos, idem. E entre os dígitos numéricos, também temos ordem crescente correspondente aos valores. Mas todos os minúsculos são "maiores" do que qualquer maiúsculo, que por sua vez são "maiores" do que qualquer dígito numérico. Símbolos, operadores e sinais de pontuação estão em posições diversas.

In [ ]:
for letra in range(255):
    print("%d = %c" %(letra, chr(letra)))

0 = 
1 = 
2 = 
3 = 
4 = 
5 = 
6 = 
7 = 
8 = 
9 = 	
10 = 

11 = 
12 = 
13 = 
14 = 
15 = 
16 = 
17 = 
18 = 
19 = 
20 = 
21 = 
22 = 
23 = 
24 = 
25 = 
26 = 
27 = 
28 = 
29 = 
30 = 
31 = 
32 =  
33 = !
34 = "
35 = #
36 = $
37 = %
38 = &
39 = '
40 = (
41 = )
42 = *
43 = +
44 = ,
45 = -
46 = .
47 = /
48 = 0
49 = 1
50 = 2
51 = 3
52 = 4
53 = 5
54 = 6
55 = 7
56 = 8
57 = 9
58 = :
59 = ;
60 = <
61 = =
62 = >
63 = ?
64 = @
65 = A
66 = B
67 = C
68 = D
69 = E
70 = F
71 = G
72 = H
73 = I
74 = J
75 = K
76 = L
77 = M
78 = N
79 = O
80 = P
81 = Q
82 = R
83 = S
84 = T
85 = U
86 = V
87 = W
88 = X
89 = Y
90 = Z
91 = [
92 = \
93 = ]
94 = ^
95 = _
96 = `
97 = a
98 = b
99 = c
100 = d
101 = e
102 = f
103 = g
104 = h
105 = i
106 = j
107 = k
108 = l
109 = m
110 = n
111 = o
112 = p
113 = q
114 = r
115 = s
116 = t
117 = u
118 = v
119 = w
120 = x
121 = y
122 = z
123 = {
124 = |
125 = }
126 = ~
127 = 
128 = €
129 = 
130 = ‚
131 = ƒ
132 = „
133 = …
134 = †
135 = ‡
136 = ˆ
137 = ‰
138 = Š

**Códigos de controle**

>0 - 31


**Caracteres imprimíveis**

> 32 - 127

**Códigos de controle**

>128 - 255

**Outras codificações:**
- EBCDIC, UTF-8, UTF-16, ISO,

##**Conversão de código**

In [ ]:
a = 'x'
b = ord(a)
print(f'O código para {a} é {b}')
c = chr(b)
print(f'A letra para {b} é {c}')

O código para x é 120
A letra para 120 é x
